applications/tari_validator_node/src/p2p/services/committee_state_sync.rs

```diff
@@ -113,7 +118,24 @@ impl CommitteeStateSync {
         &self,
         committee_vns: Committee<CommsPublicKey>,
         shard_range: RangeInclusive<ShardId>,
-    ) -> Result<(), CommitteeStateSyncError> {
+        epoch: Epoch,
+    ) -> Result<Vec<(BlockId, BlockId)>, CommitteeStateSyncError> {
+        let tip = self.shard_store.with_read_tx(|tx| Block::get_tip(tx, epoch))?;
+        // Missing children is an array (from, to), where we have `from` and `to` blocks, but nothing in between. In
+        // case the `to` is `None`, then we want everything up to the tip.
+
+        let mut missing_blocks = {
+            let mut tx = self.global_db.create_transaction()?;
+            match self
+                .global_db
+                .metadata(&mut tx)
+                .get_metadata::<Vec<(BlockId, BlockId)>>(MetadataKey::MissingBlocks)?
```

Comment:
I don't really understand why we need to persist this data.

Reply:
This is for the case when you sync fails in the middle. So you can end up with a ranges of blocks.
E.g. you have tip 10, and now blocks 20 comes in. You start syncing (the blocks are coming from the tip). You will receive 20,19,18 (and it fails). So now you have 1-10,18-20. You start syncing again and now the tip is 30, so you get 30down to 25 (and fail again). This way you can end up with many block ranges missing. And we probably don't want to sync all the time from the tip to our last know tip. In this example you don't want again all blocks from 10 to 30. But I'm open to any other solution.